### DISCUSSION TITLE
CaptumExplainer for heterogeneous graphs

### DISCUSSION THREAD

**Discussion by Faiail** [pyg-team/pytorch_geometric#9256] [Q&A]
Hello,
I'm trying to run `Captum Explainer` for a node classification task. However, I'm facing some issues.
First, unused parameters have been detected (maybe some nodes have not been reached for a limited number of hops during the message passing), but I've forced the parameter in the code.
However, I've still `None` gradients, so I couldn't find a feasible solution.

To get some insights in the code, I'm posting the model:
![image](https://github.com/pyg-team/pytorch_geometric/assets/46848656/09067f59-7bf5-4be7-b851-20ee4fab3bb2)

The explainer setting:
![image](https://github.com/pyg-team/pytorch_geometric/assets/46848656/a93694bf-af19-4904-b129-4d884237cd18)

and the error
![image](https://github.com/pyg-team/pytorch_geometric/assets/46848656/e8954d63-c56d-454b-9795-8a0e97f6a1b8)

Thanks a lot for the help!!

**Comment by Faiail** [comment]
**Update**: I've tried with a dummy model with more GNN layers and everything works. Thus, it is an issue related to unreached nodes during the message-passing phase. How can I solve that?

**Reply by rusty1s** [reply]
Let me take a look if we can fix this on our side. In your case, you could simply remove the node/edge types which are not reachable within the number of message passing hops, e.g.:
```python
del data[node_type]
del data[edge_type]

**Reply by Faiail** [reply]
Hello,
Thanks for your help.
However, I have found a workable solution, which could be directly integrated in the framework.
I have extended the IntegratedGradients class, which solves the problem of "None" gradients (problem that is related to the fact that no hook is registered because that node does not take part to the computational graph), by simply doing (in the _attribute method):
```
grads = [
            (
                grad
                if grad is not None
                else torch.zeros_like(inputs[ix])
            )
            for (ix, grad) in enumerate(grads)
        ]
```
In this way, since the nodes are not taken into account in the forward pass, their mask would theoretcally take 0 importance.


Moreover, creating a new class, I needed also to extend the CaptumExplainer class, by adding this new IG engine, doing:

```
from torch_geometric.explain import CaptumExplainer


class myCaptumExplainer(CaptumExplainer):
    SUPPORTED_METHODS = CaptumExplainer.SUPPORTED_METHODS
    SUPPORTED_METHODS.append("GraphIntegratedGradients") # new class to be included
```

Last but not least, there is still the problem of allowing unused parameters in autograd. Please add a parameter to the explainer, which is forwarded to that part of code.

Thanks.

**Reply by rusty1s** [reply]
> Last but not least, there is still the problem of allowing unused parameters in autograd. Please add a parameter to the explainer, which is forwarded to that part of code.

Do you mind clarifying? To which part of the code are you referring to?
